packages/alfa-rules/src/sia-r69/rule.ts

```diff
@@ -0,0 +1,388 @@
+import { Rule } from "@siteimprove/alfa-act";
+import { Role } from "@siteimprove/alfa-aria";
+import { RGB, Percentage, Current, System } from "@siteimprove/alfa-css";
+import { Device } from "@siteimprove/alfa-device";
+import { Element, Text, Namespace, Node } from "@siteimprove/alfa-dom";
+import { Iterable } from "@siteimprove/alfa-iterable";
+import { Option, None } from "@siteimprove/alfa-option";
+import { Predicate } from "@siteimprove/alfa-predicate";
+import { Ok, Err } from "@siteimprove/alfa-result";
+import { Style } from "@siteimprove/alfa-style";
+import { Page } from "@siteimprove/alfa-web";
+
+import { expectation } from "../common/expectation";
+
+import { hasCategory } from "../common/predicate/has-category";
+import { hasRole } from "../common/predicate/has-role";
+import { isDisabled } from "../common/predicate/is-disabled";
+import { isIgnored } from "../common/predicate/is-ignored";
+import { isVisible } from "../common/predicate/is-visible";
+
+import { Question } from "../common/question";
+
+const { reduce, some, flatMap, map, concat } = Iterable;
+const { and, not, equals, test } = Predicate;
+const { min, max } = Math;
+
+export default Rule.Atomic.of<Page, Text, Question>({
+  uri: "https://siteimprove.github.io/sanshikan/rules/sia-r69.html",
+  evaluate({ device, document }) {
+    return {
+      applicability() {
+        return visit(document);
+
+        function* visit(node: Node): Iterable<Text> {
+          if (Element.isElement(node)) {
+            if (
+              test(Element.hasNamespace(not(equals(Namespace.HTML))), node) ||
+              test(hasRole(hasCategory(equals(Role.Category.Widget))), node) ||
+              test(
+                and(hasRole(Role.hasName(equals("group"))), isDisabled),
+                node
+              )
+            ) {
+              return;
+            }
+          }
+
+          if (Text.isText(node)) {
+            if (test(and(isVisible(device), not(isIgnored(device))), node)) {
+              yield node;
+            }
+          }
+
+          const children = node.children({ flattened: true, nested: true });
+
+          for (const child of children) {
+            yield* visit(child);
+          }
+        }
+      },
+
+      expectations(target) {
+        const foregrounds = Question.of(
+          "foreground-colors",
+          "color[]",
+          target,
+          "What are the foreground colors of the text node?"
+        );
+
+        const backgrounds = Question.of(
+          "background-colors",
+          "color[]",
+          target,
+          "What are the background colors of the text node?"
+        );
```

Comment:
I kinda fail to see how these questions are not ask every single time 🤔 
But it seem to work 🤷

Reply:
The magic of monads 😄 The type of the first `result` is basically `Question<Question<..>>` with the outer `Question` providing the foreground colours and the inner `Question` providing the background colours:

https://github.com/Siteimprove/alfa/blob/4669b7900bc4d5a3e716a00879cc42516238df08/packages/alfa-rules/src/sia-r69/rule.ts#L77-L97

We then get attempt to determine the foreground colours automatically and, if that succeeds, we use them to unwrap a single layer of `result` so that it becomes `Question<...>`:

https://github.com/Siteimprove/alfa/blob/4669b7900bc4d5a3e716a00879cc42516238df08/packages/alfa-rules/src/sia-r69/rule.ts#L102-L103

We then repeat the same thing for the background colours to unwrap the last layer of the _new_ `result`:

https://github.com/Siteimprove/alfa/blob/4669b7900bc4d5a3e716a00879cc42516238df08/packages/alfa-rules/src/sia-r69/rule.ts#L104-L108

That's also the reason for naming both variables `result`. If all else fails, we simply fall back to the old `result`:

https://github.com/Siteimprove/alfa/blob/4669b7900bc4d5a3e716a00879cc42516238df08/packages/alfa-rules/src/sia-r69/rule.ts#L109

The result, no pun intended, of all this is that we either end up with an entirely unwrapped `result`, a single `Question<...>` layer, or a double `Question<Question<...>>` layer.

Reply:
I think I misunderstood l103/106 which actually means "Hey, I got some answer to your question called `result` and I'm giving it now"…
Make sense now.